File: python/plugins/wrap_code_plugin.py
import json
import logging
import os
from pathlib import Path
from typing import Any, Dict, Union

from argument_file_utils import complete_arguments_processing, merge_and_canonize_argument_file
from cli_arguments_utils import CliArgDataObject
from models.document import Document
from models.options import Options
from models.page_metadata_handlers import PageMetadataHandlers
from output_utils import output_page, MARKDOWN
from plugins.md2html_plugin import Md2HtmlPlugin
from plugins.plugin_utils import dict_from_string_or_object
from utils import read_lines_from_cached_file, relativize_relative_resource, UserError

logger = logging.getLogger(__name__)

MODULE_DIR = Path(__file__).resolve().parent


class WrapCodeData:
    def __init__(self):
        self.style = ""
        self.variables = {}
        self.document_dict: dict = {}
        self.document_obj: Document = Document()


class WrapCodePlugin(Md2HtmlPlugin):

    def __init__(self):
        super().__init__()
        self.data: Dict[str, WrapCodeData] = {}
        self.processed_cache: Dict[str, str] = {}
        self.plugins_for_output = []
        self.app_options = None
        self.dry_run = False
        with open(MODULE_DIR.joinpath('wrap_code_metadata_schema.json'), 'r',
                  encoding="utf-8") as schema_file:
            self.metadata_schema = json.load(schema_file)

    def accept_data(self, data):
        self.assure_accept_data_once()
        self.validate_data_with_file(data, MODULE_DIR.joinpath('wrap_code_schema.json'))

        for marker, data_dict in data.items():
            wrapped_document_data = WrapCodeData()
            wrapped_document_data.document_dict = {k: v for k, v in data_dict.items()}
            wrapped_document_data.document_dict.pop('style', None)
            wrapped_document_data.document_dict.pop('variables', None)
            wrapped_document_data.style = data_dict.get("style", wrapped_document_data.style)
            wrapped_document_data.variables = data_dict.get("variables",
                                                            wrapped_document_data.variables)
            self.data[marker.upper()] = wrapped_document_data

    def pre_initialize(self, argument_file: dict, cli_args: CliArgDataObject,
                       plugins: list) -> Dict[str, Any]:

        for marker_data in self.data.values():
            this_argument_file = argument_file.copy()
            this_document_dict = marker_data.document_dict.copy()
            this_document_dict["input"] = "fictional"
            this_document_dict["output"] = "fictional"
            this_argument_file['documents'] = [this_document_dict]

            canonized_argument_file = merge_and_canonize_argument_file(this_argument_file, cli_args)
            arguments, _ = complete_arguments_processing(canonized_argument_file, plugins)

            document_obj = arguments.documents[0]
            document_obj.input_file = document_obj.input_file[:-9]
            document_obj.output_file = document_obj.output_file[:-9]
            marker_data.document_obj = document_obj

        return {}

    def is_blank(self) -> bool:
        return not bool(self.data)

    def accept_app_data(self, plugins: list, options: Options,
                        metadata_handlers: PageMetadataHandlers):
        self.plugins_for_output = plugins
        self.app_options = options

    def page_metadata_handlers(self):
        return [(self, marker, False) for marker in self.data.keys()]

    def accept_page_metadata(self, doc: Document, marker: str, metadata_str: str, metadata_section,
                             visited_markers: Union[Dict[str, None]] = None):
        marker = marker.upper()
        marker_data = self.data[marker]
        document_obj = marker_data.document_obj

        try:
            metadata = dict_from_string_or_object(metadata_str.strip(), "file",
                                                  self.metadata_schema)
        except UserError as e:
            raise UserError(f"Error in inclusion: {str(e)}")

        file_path = metadata["file"]
        style = metadata.get("style", marker_data.style)

        input_file = Path(document_obj.input_file).joinpath(file_path)
        input_file_str = str(input_file).replace("\\", "/")
        cache_key = marker + "|" + input_file_str
        output_file_str = self.processed_cache.get(cache_key)
        if not output_file_str:
            output_file = Path(document_obj.output_file).joinpath(file_path + ".html")
            output_file_str = str(output_file).replace("\\", "/")

            need_to_generate = True
            if not document_obj.force and output_file.exists():
                output_file_mtime = os.path.getmtime(output_file)
                input_file_mtime = os.path.getmtime(input_file)
                if output_file_mtime > input_file_mtime:
                    logger.info('Wrapped output file is up-to-date. Skipping: %s', output_file_str)
                    need_to_generate = False

            if need_to_generate and not self.dry_run:
                document_obj = document_obj.copy(input_file=input_file_str,
                                                 output_file=output_file_str)

                try:
                    content = read_lines_from_cached_file(document_obj.input_file)
                except FileNotFoundError as e:
                    raise UserError(
                        f"Error processing page metadata block: {type(e).__name__}: {e}")
                doc_content = ("````" + style + "\n" + content + "\n" + "````")
                substitutions = {'content': MARKDOWN.convert(source=doc_content)}

                variables = marker_data.variables.copy()
                file_name = str(Path(file_path).name)
                variables.update({"title": file_name, "wrap_code_path": file_path,
                                  "wrap_code_file_name": file_name})

                output_page(document_obj, self.plugins_for_output, substitutions, variables)

                logger.info('Wrapped output file generated: %s', document_obj.output_file)

            self.processed_cache[cache_key] = output_file_str

        return relativize_relative_resource(output_file_str, doc.output_file)

File: java/src/main/java/world/md2html/plugins/WrapCodePlugin.java
package world.md2html.plugins;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.ValueNode;
import world.md2html.Md2Html;
import world.md2html.options.argfile.ArgFileParseException;
import world.md2html.options.model.ArgFile;
import world.md2html.options.model.CliOptions;
import world.md2html.options.model.Document;
import world.md2html.options.model.SessionOptions;
import world.md2html.options.model.raw.ArgFileDocumentRaw;
import world.md2html.options.model.raw.ArgFileRaw;
import world.md2html.pagemetadata.PageMetadataHandlersWrapper;
import world.md2html.utils.CheckedIllegalArgumentException;
import world.md2html.utils.Utils;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.attribute.FileTime;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static world.md2html.Md2HtmlUtils.generateHtml;
import static world.md2html.options.argfile.ArgFileParsingHelper.completeArgFileProcessing;
import static world.md2html.options.argfile.ArgFileParsingHelper.mergeAndCanonizeArgFileRaw;
import static world.md2html.utils.JsonUtils.OBJECT_MAPPER_FOR_BUILDERS;
import static world.md2html.utils.JsonUtils.deJson;
import static world.md2html.utils.Utils.getCachedString;
import static world.md2html.utils.Utils.relativizeRelativeResource;
import static world.md2html.utils.Utils.supplyWithFileExceptionAsUserError;

public class WrapCodePlugin extends AbstractMd2HtmlPlugin implements PageMetadataHandler {

    private static class WrapCodeData {
        private String style = "";
        private Map<String, Object> variables = new HashMap<>();
        private ObjectNode documentJson;
        private Document documentObj;
    }

    private Map<String, WrapCodeData> data;
    private final Map<String, String> processedCache = new HashMap<>();

    public void setDryRun(boolean dryRun) {
        this.dryRun = dryRun;
    }

    private boolean dryRun;

    private SessionOptions options;
    private List<Md2HtmlPlugin> plugins;

    @Override
    public void acceptData(JsonNode data) throws ArgFileParseException {

        assureAcceptDataOnce();
        validateInputDataAgainstSchemaFromResource(data, "plugins/wrap_code_schema.json");

        Map<String, WrapCodeData> dataMap = new LinkedHashMap<>();
        Iterator<Map.Entry<String, JsonNode>> fieldIterator = data.fields();
        while (fieldIterator.hasNext()) {
            Map.Entry<String, JsonNode> fieldEntry = fieldIterator.next();
            ObjectNode markerDataNode = fieldEntry.getValue().deepCopy();

            WrapCodeData wrapCodeData = new WrapCodeData();
            wrapCodeData.documentJson = markerDataNode;

            ValueNode styleNode = (ValueNode) markerDataNode.get("style");
            if (styleNode != null) {
                wrapCodeData.style = styleNode.asText();
            }
            ObjectNode variablesNode = (ObjectNode) markerDataNode.get("variables");
            if (variablesNode != null) {
                //noinspection unchecked
                wrapCodeData.variables = (Map<String, Object>) deJson(variablesNode);
            }
            markerDataNode.remove("style");
            markerDataNode.remove("variables");
            dataMap.put(fieldEntry.getKey().toUpperCase(), wrapCodeData);
        }

        this.data = dataMap;
    }

    @Override
    public boolean isBlank() {
        return data.isEmpty();
    }

    @Override
    public Map<String, JsonNode> preInitialize(ArgFileRaw argFileRaw, CliOptions cliOptions,
                                               Map<String, Md2HtmlPlugin> plugins) {

        for (WrapCodeData wrapCodeData : data.values()) {

            ArgFileDocumentRaw argFileDocumentRaw;
            try {
                argFileDocumentRaw = OBJECT_MAPPER_FOR_BUILDERS
                        .treeToValue(wrapCodeData.documentJson, ArgFileDocumentRaw.class);
            } catch (JsonProcessingException e) {
                throw new RuntimeException(e);
            }

            ArgFileRaw newArgFileRaw = argFileRaw.toBuilder().documents(Collections.singletonList(
                    argFileDocumentRaw.toBuilder()
                            .input("fictional")
                            .output("fictional")
                            .build())
            ).build();

            ArgFileRaw canonizedArgFileRaw =
                    mergeAndCanonizeArgFileRaw(newArgFileRaw, cliOptions);
            ArgFile arguments =
                    completeArgFileProcessing(canonizedArgFileRaw, plugins).getValue0();

            Document documentObj = arguments.getDocuments().get(0);
            String newInput = documentObj.getInput();
            String newOutput = documentObj.getOutput();
            wrapCodeData.documentObj = documentObj.toBuilder()
                    .input(newInput.substring(0, newInput.length() - 9))
                    .output(newOutput.substring(0, newOutput.length() - 9))
                    .build();
        }
        return Collections.emptyMap();
    }

    @Override
    public void acceptAppData(SessionOptions options, List<Md2HtmlPlugin> plugins,
                              PageMetadataHandlersWrapper metadataHandlers) {
        this.options = options;
        this.plugins = plugins;
    }

    @Override
    public List<PageMetadataHandlerInfo> pageMetadataHandlers() {
        return this.data.keySet().stream().map(marker ->
                new PageMetadataHandlerInfo(this, marker, false)).collect(Collectors.toList());
    }

    @Override
    public Map<String, Object> variables(Document document) {
        return Collections.emptyMap();
    }

    @Override
    public String acceptPageMetadata(Document document, String marker, String metadata,
                                     String metadataSection) throws PageMetadataException {

        marker = marker.toUpperCase();
        WrapCodeData markerData = this.data.get(marker);
        metadata = metadata.trim();
        Document documentObj = markerData.documentObj;

        Path inputFile = Paths.get(documentObj.getInput(), metadata);
        String inputFileStr = inputFile.toString().replace("\\", "/");
        String cacheKey = marker + "|" + inputFileStr;

        String outputFileStr = this.processedCache.get(cacheKey);
        if (outputFileStr == null) {
            Path outputFile = Paths.get(documentObj.getOutput(), metadata + ".html");
            outputFileStr = outputFile.toString().replace("\\", "/");

            boolean needToGenerate = true;
            if (!documentObj.isForce() && Files.exists(outputFile)) {
                FileTime inputFileTime;
                FileTime outputFileTime;
                try {
                    inputFileTime = Files.getLastModifiedTime(inputFile);
                    outputFileTime = Files.getLastModifiedTime(outputFile);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
                if (outputFileTime.compareTo(inputFileTime) > 0) {
                    if (document.isVerbose()) {
                        System.out.println("Wrapped output file is up-to-date. Skipping: "
                                + document.getOutput());
                        needToGenerate = false;
                    }
                }
            }

            if (needToGenerate && !this.dryRun) {
                documentObj = documentObj.toBuilder()
                        .input(inputFileStr)
                        .output(outputFileStr)
                        .build();

                String content = supplyWithFileExceptionAsUserError(
                        () -> getCachedString(inputFile, Utils::readStringFromUtf8File),
                        "Error processing page metadata block"
                );
                String docContent = "````" + markerData.style + "\n" +
                         content + "\n" + "````";

                Map<String, Object> substitutions = new HashMap<>();
                substitutions.put("content", generateHtml(docContent));

                Map<String, Object> variables = new HashMap<>(markerData.variables);
                String fileName = Paths.get(metadata).getFileName().toString();
                variables.put("title", fileName);
                variables.put("wrap_code_path", metadata);
                variables.put("wrap_code_file_name", fileName);

                Md2Html.outputPage(documentObj, this.plugins, substitutions, this.options,
                        variables);

                if (documentObj.isVerbose()) {
                    System.out.println("Wrapped output file generated: " + documentObj.getOutput());
                }
                if (documentObj.isReport()) {
                    System.out.println(documentObj.getOutput());
                }
            }
            this.processedCache.put(cacheKey, outputFileStr);
        }

        try {
            return relativizeRelativeResource(outputFileStr, document.getOutput());
        } catch (CheckedIllegalArgumentException e) {
            throw new RuntimeException(e);
        }
    }
}
